src/fxos/readGeckoCommit.js

```diff
@@ -0,0 +1,97 @@
+let assert = require('assert');
+let debug = console.log.bind(console, '[fxos/readGeckoCommit]');
+let denodeify = require('denodeify');
+let fs = require('mz/fs');
+let path = require('path');
+let sax = require('sax');
+let tmpdir = denodeify(require('tmp').dir);
+
+const SOURCES_XML = '/system/sources.xml';
+const APPLICATION_INI = '/system/b2g/application.ini';
+const PLATFORM_INI = '/system/b2g/platform.ini';
+
+function pullIniFile(adb, local, file) {
+  return adb
+    .pull(file, local)
+    .then(() => path.join(local, path.basename(file)));
+}
+
+function readGeckoRevisionFromIni(iniFile) {
+  let revision;
+
+  return new Promise((resolve, reject) => {
+    let stream = fs.createReadStream(iniFile, {encoding: 'utf8'});
+
+    // The Gecko revision will be assigned to key named SourceStamp, in the
+    // line format:  SourceStamp=${REVISION}
+    stream.on('data', buffer => {
+      let line = buffer
+        .split('\n')
+        .find(line => line.includes('SourceStamp'));
+
+      if (!line) {
+        return;
+      }
+
+      revision = line.split('=')[1];
+      resolve(revision);
+    });
+
+    stream.on('end', () => {
+      if (!revision) {
+        reject('Unable to find Gecko revision in INI file');
+      }
+    });
+  });
+}
+
+function getGeckoRevisionFromIni(adb, local) {
+  return Promise
+    .resolve()
+    .then(() => pullIniFile(adb, local, APPLICATION_INI))
+    .catch(() => pullIniFile(adb, local, PLATFORM_INI))
+    .then(file => readGeckoRevisionFromIni(file));
+}
+
+function getGeckoRevisionFromXml(local) {
+  let revision;
+
+  return new Promise((resolve, reject) => {
+    let stream = sax.createStream();
+
+    // The Gecko revision lives in a tag that MAY be HTML-commented
+    stream.on('opentag', node => {
```

Comment:
I'd agree with you on text streams where the data comes from the `data` event, but with object-based streams like this one, the data comes from the `opentag` event and I don't see any real easy way to abstract it the same way. I'll chew on it to see if I can think of anything.